### PR TITLE
Fix up Absolute value #abs

### PR DESCRIPTION
**Pull request by chick**
Defines .abs() in Num.  Implements Issue #466 
Implement .abs() in UInt
Change .abs() in SInt to return an SInt
Change .abs() in FixedPoint to return a FixedPoint
Added  a couple of tests
Add some scala style suppression to Bits so I can read code in IntelliJ

### CODE REVIEW

File: chiselFrontend/src/main/scala/chisel3/core/Bits.scala [hunk 1]
```diff
@@ -905,10 +916,8 @@ sealed class FixedPoint private (width: Width, val binaryPoint: BinaryPoint, lit
   def do_=/= (that: FixedPoint)(implicit sourceInfo: SourceInfo): Bool = compop(sourceInfo, NotEqualOp, that)
   def do_=== (that: FixedPoint)(implicit sourceInfo: SourceInfo): Bool = compop(sourceInfo, EqualOp, that)
 
-  final def abs(): UInt = macro SourceInfoTransform.noArg
-
-  def do_abs(implicit sourceInfo: SourceInfo): UInt = {
-    Mux(this < FixedPoint.fromBigInt(0), (FixedPoint.fromBigInt(0)-this).asUInt, this.asUInt)
+  def do_abs(implicit sourceInfo: SourceInfo): FixedPoint = {
+    Mux(this < FixedPoint.fromBigInt(0), (FixedPoint.fromBigInt(0)-this), this)
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Bits.scala]: Any reason you can't use 0.F here?

File: src/test/scala/chiselTests/FixedPointSpec.scala [hunk 2]
```diff
@@ -21,23 +21,35 @@ class FixedPointLiteralSpec extends FlatSpec with Matchers {
...
+  val negativefp = -3.5.F(binaryPoint = 4)
+  val positivefp = 3.5.F(binaryPoint = 4)
+
+  assert(uint2fp === uint_result)
+  assert(sint2fp === sint_result)
+  assert(fp2fp   === fp_result)
+
+  assert(uint.abs() === uint)
+  assert(sint.abs() === sintpos)
+  assert(sint.abs() =/= wrongSIntPos)
```
Comment (ducky64) [inline src/test/scala/chiselTests/FixedPointSpec.scala]: Can you add tests to ensure sintpos.abs and positivefp.abs also work?

File: src/test/scala/chiselTests/FixedPointSpec.scala [hunk 3]
```diff
@@ -21,23 +21,35 @@ class FixedPointLiteralSpec extends FlatSpec with Matchers {
...
-    val uint2fp = fp_tpe.fromBits(uint)
-    val sint2fp = fp_tpe.fromBits(sint)
-    val fp2fp   = fp_tpe.fromBits(fp)
-
-    assert(uint2fp === uint_result)
-    assert(sint2fp === sint_result)
-    assert(fp2fp   === fp_result)
-
-    stop()
+  val uint = 3.U(4.W)
```
Comment (ducky64) [inline src/test/scala/chiselTests/FixedPointSpec.scala]: Shouldn't the integer stuff go into its own test suite? Not the "FixedPointFromBitsTester"?

Comment (chick) [inline src/test/scala/chiselTests/FixedPointSpec.scala]: done

**Comment by chick** [comment]
@ducky64 comments addressed

**Comment by ucbjrl** [comment]
Unfortunately, this appears to break rocket-chip:
[jenkins build 1121](http://a17.millennium.berkeley.edu/jenkins/job/rocket-chip_with_chisel3/1121/)
```
[error] /vm/home/jenkins/workspace/rocket-chip_with_chisel3/src/main/scala/util/Misc.scala:101: type mismatch;
[error]  found   : Chisel.UInt
[error]     (which expands to)  chisel3.core.UInt
[error]  required: chisel3.core.SInt
[error]       var s = digit(q % rad)
[error]                         ^
[error] /vm/home/jenkins/workspace/rocket-chip_with_chisel3/src/main/scala/util/Misc.scala:104: type mismatch;
[error]  found   : Chisel.UInt
[error]     (which expands to)  chisel3.core.UInt
[error]  required: chisel3.core.SInt
[error]         q = q / rad
[error]                 ^
[error] /vm/home/jenkins/workspace/rocket-chip_with_chisel3/src/main/scala/util/Misc.scala:105: type mismatch;
[error]  found   : Chisel.UInt
[error]     (which expands to)  chisel3.core.UInt
[error]  required: chisel3.core.SInt
[error]         val placeSpace = q === UInt(0)
[error]                                    ^
[error] /vm/home/jenkins/workspace/rocket-chip_with_chisel3/src/main/scala/util/Misc.scala:108: type mismatch;
[error]  found   : Chisel.UInt
[error]     (which expands to)  chisel3.core.UInt
[error]  required: chisel3.core.SInt
[error]         s = Cat(Mux(placeSpace, space, digit(q % rad)), s)
[error]                                                  ^
[warn] one warning found
[error] four errors found
[error] (rocketchip/compile:compileIncremental) Compilation failed
```

**Comment by ucbjrl** [comment]
I'll put together a PR for rocket-chip.

**Comment by chick** [comment]
@ucbjrl thanks, you can test it better than I.

**Pull request merged**